### DISCUSSION TITLE
Error: GLB must have 0–1 buffers. when writing glb

### DISCUSSION THREAD

**Discussion by daniel-gato** [donmccurdy/glTF-Transform#555] [Q&A] [answered]
I'm trying to convert a gltf to glb.

The gltf has this specificities:
- uses draco
- uses texture transform
- the textures and binaries are remote urls
- the gltf references many .bin and .jpg 

I keep getting this message:
- Error: GLB must have 0–1 buffers.

I don't have control over that gltf and it displays correctly on gltf viewers. 

I'm not sure why this happens. I would expect the write() to include all .bins into the file.

```typescript
import fetch from 'node-fetch';
import { NodeIO } from '@gltf-transform/core';
import { KHRONOS_EXTENSIONS } from '@gltf-transform/extensions';

const io = new NodeIO(fetch)
  .setAllowHTTP(true)
  .registerExtensions(KHRONOS_EXTENSIONS);

const doc = await io.read('/tmp/model.gltf');
await io.write('/tmp/model.glb', doc);
```

Am I missing something here?

**Comment by donmccurdy** [comment, marked as answer]
Hi @daniel-gato! By convention a `.glb` is usually self-contained (i.e. does not reference external `.bin` files) and so any buffers need to be consolidated if there are >1. The I/O classes don't currently do that, in order to avoid mutating the original Document during export unintentionally.

If you add the following code before writing the Document, it will consolidate any extra buffers:

```js
const buffer = doc.getRoot().listBuffers()[0];
doc.getRoot().listAccessors()
  .forEach((a) => a.setBuffer(buffer));
doc.getRoot().listBuffers()
  .forEach((b, index) => (index > 0 ? b.dispose() : null));
```

EDIT: In more recent versions of the library, this shortcut can be used instead:

```js
import { unpartition } from '@gltf-transform/functions';

await document.transform(unpartition());
```

**Comment by daniel-gato** [comment]
I’m not sure to understand the code. Will that remove the extra buffers or merge them into one? I’m other terms, will the resulting gob be visually the same than the initial gltf?

My understanding was that a gltf contains the binary geometry data into one or several external binary files. While the glob contains all those files and the gltf structure merged into one single binary file.

**Reply by donmccurdy** [reply]
Each Buffer is a flexible container for binary data, like mesh geometry or animation. Any number of vertex attributes or animation tracks can be stored in a single Buffer. To summarize what the code is doing:

```js
// select the first buffer
const buffer = doc.getRoot().listBuffers()[0];

// reassign all binary data to the first buffer
doc.getRoot().listAccessors()
  .forEach((a) => a.setBuffer(buffer));

// delete all buffers except the first
doc.getRoot().listBuffers()
  .forEach((b, index) => (index > 0 ? b.dispose() : null));
```

There will be no visual differences in the GLB, but all of the external `.bin` (buffer) files will be stored in the file's single internal buffer. Only one buffer can be embedded in a `.glb`, so we have to merge their data.

**Reply by mnakaji1-avail** [reply]
I'm trying to load a gltf in json format and generate a glb document. I read a json file and some asset files and generated a JSONDocument. I converted it to a Document in WebIO.readJSON(), and as you can see in this thread, I was able to generate a Document with unpartition() without any errors while loading, but I have a problem. The Document I created is an error in gltf varidation. There are more than 500 errors for ACCESSOR_VECTOR3_NON_UNIT. However, it doesn't seem to have any visual problems. I would like to avoid errors if possible, but what should I do? Thank you for your kindness.

**Reply by donmccurdy** [reply]
@mnakaji1-avail  would you mind starting a [new discussion](https://github.com/donmccurdy/glTF-Transform/discussions/new/choose) for this question?
